android-core/src/main/java/com/mparticle/MParticle.java

```diff
@@ -261,6 +281,38 @@ public static void setInstance(@Nullable MParticle instance) {
         MParticle.instance = instance;
     }
 
+    /**
+     * Switch the SDK to a new API key and secret.
+     * Will first batch all events that have not been sent to mParticle into upload records,
+     * then all SDK state including user defaults, database (except uploads), etc will be
+     * completely reset. After that, {@link #start(MParticleOptions)} )} will be called with
+     * the new key and secret and the SDK will initialize again as if it is a new app launch.
+     *
+     @param options Required to initialize the SDK properly
+     */
+    public static void switchWorkspace(@NonNull MParticleOptions options) {
+        if (instance != null) {
+            // End session if active
+            if (instance.isSessionActive()) {
```

Comment:
who is to say that instance didn't become null due to another thread in between 294 and 296? 

i would:
- synchronize on the class (see getinstance) so nobody crazy can call this method simultaneously and
- grab a local reference to the instance, and then check if that's null and use that.

i'd lock on the instance (eg `synchronized (MParticle.class))` and to be extra safe

Reply:
Lock added